File: native/ios/SmartMapsAE/Core/AECore.swift
// AECore.swift — the A/E All-In-One orchestrator.
//
// Owns every engine and drives one dynamic tick loop. Engines only ever meet
// here; they never call each other. Modules activate purely by having data —
// no modes, no configuration, no fallback logic. SwiftUI observes the
// @Published state; it never reaches into an engine. Mirrors the Kotlin
// AECore tick pipeline exactly:
//
//   1. movement  ← IMU window (+ GNSS speed corroboration)
//   2. fused fix ← inverse-variance fusion of all fresh sources, IMU-gated
//   3. snap      ← road graph projection with hysteretic lock
//   4. DR        ← when no fresh fix: advance along the snapped edge
//   5. route     ← live remaining distance / ETA / next instruction
//   6. transit   ← provider board refresh around the fused position

import Foundation
import Combine

public final class AECore: ObservableObject {

    public let fusion = LocationFusionEngine()
    public let movement = MovementEngine()
    public let spatial = SpatialEngine()
    public let deadReckoning = DeadReckoningEngine()
    public let routing = RoutingEngine()
    public let transport = TransportTimeEngine()
    public let cache: OfflineGeometryCache

    // ── observable state (UI binds to these) ────────────────────────────────
    @Published public private(set) var fix: FusedFix?
    @Published public private(set) var snap: SnapResult?
    @Published public private(set) var movementState = MovementEstimate(
        state: .unknown, confidence: 0, imuActivity: 0, timestampMs: 0
    )
    @Published public private(set) var route: Route?
    @Published public private(set) var nextStep: RouteStep?
    @Published public private(set) var remainingM = 0.0
    @Published public private(set) var remainingEtaS = 0.0
    @Published public private(set) var arrivals: [TransitArrival] = []

    private var timer: Timer?
    private var lastTickMs: Int64 = 0
    private var lastGoodFixMs: Int64 = 0
    private let tickInterval: TimeInterval
    private let now: () -> Int64

    public init(
        cache: OfflineGeometryCache = OfflineGeometryCache(),
        tickInterval: TimeInterval = 0.2,
        now: @escaping () -> Int64 = { Int64(Date().timeIntervalSince1970 * 1000) }
    ) {
        self.cache = cache
        self.tickInterval = tickInterval
        self.now = now
    }

    // ── lifecycle ───────────────────────────────────────────────────────────

    public func start(graphKey: String = "default") {
        guard timer == nil else { return }
        if let graph = cache.load(key: graphKey) {
            spatial.setGraph(graph)
        }
        lastTickMs = now()
        timer = Timer.scheduledTimer(withTimeInterval: tickInterval, repeats: true) { [weak self] _ in
            self?.tick()
        }
    }

    public func stop() {
        timer?.invalidate()
        timer = nil
    }

    /// Install a road graph (e.g. freshly downloaded) — write-through cache.
    public func installGraph(key: String, graph: RoadGraph) {
        cache.save(key: key, graph: graph)
        spatial.setGraph(graph)
    }

    // ── sensor feeds (SensorHub pushes here) ────────────────────────────────

    public func pushFix(_ raw: RawFix) { fusion.ingest(raw) }

    public func pushMotion(_ sample: MotionSample) { movement.push(sample) }

    // ── routing ─────────────────────────────────────────────────────────────

    @discardableResult
    public func navigateTo(_ destination: GeoPoint, fromOverride: GeoPoint? = nil) -> Route? {
        guard let from = fromOverride ?? snap?.point ?? fix?.point else { return nil }
        let computed = routing.route(graph: spatial.graph(), from: from, to: destination)
        route = computed
        return computed
    }

    public func clearRoute() {
        route = nil
        nextStep = nil
        remainingM = 0
        remainingEtaS = 0
    }

    // ── the tick ────────────────────────────────────────────────────────────

    private func tick() {
        let nowMs = now()
        let dtMs = min(max(nowMs - lastTickMs, 1), 2000)
        lastTickMs = nowMs

        // 1. movement from IMU (+ GNSS speed corroboration)
        let gnssSpeed: Double? = {
            guard let f = fix, nowMs - f.timestampMs < 3000 else { return nil }
            return f.speedMps
        }()
        let move = movement.update(nowMs: nowMs, gnssSpeedMps: gnssSpeed, dtMs: dtMs)
        movementState = move

        // 2. fuse all fresh location sources
        let fused = fusion.fuse(nowMs: nowMs, movement: move, dtMs: dtMs)

        // 3–4. snap to road geometry, or dead-reckon through the outage
        if let fused {
            let snapped = spatial.snap(fused, dtMs: dtMs)
            if !fused.sources.isEmpty && fused.timestampMs > lastGoodFixMs {
                lastGoodFixMs = fused.timestampMs
            }
            let stale = nowMs - lastGoodFixMs > 3000
            if !stale, let snapped {
                snap = snapped
                var updated = fused
                updated.point = snapped.point
                updated.headingDeg = spatial.alignedHeading(fix: fused, snap: snapped)
                fix = updated
                deadReckoning.seed(speed: fused.speedMps, heading: fused.headingDeg)
            } else if let dr = deadReckoning.advance(
                spatial: spatial, lastSnap: snap, imuActivity: move.imuActivity, dtMs: dtMs
            ) {
                snap = dr
                fix = FusedFix(
                    point: dr.point,
                    accuracyM: (fix?.accuracyM ?? 20) + 0.5 * Double(dtMs) / 1000.0,
                    speedMps: deadReckoning.currentSpeedMps(),
                    headingDeg: dr.headingDeg,
                    sources: [.deadReckoning],
                    timestampMs: nowMs
                )
            } else if !fused.sources.isEmpty {
                fix = fused
                if let snapped { snap = snapped }
            }
        }

        // 5. live route progress + automatic arrival
        if let activeRoute = route, let position = fix?.point {
            let rem = routing.remaining(route: activeRoute, at: position)
            remainingM = rem.distM
            remainingEtaS = rem.etaS
            nextStep = routing.nextStep(route: activeRoute, at: position)
            if let end = activeRoute.points.last,
               haversineM(position, end) < 25, remainingM < 40 {
                clearRoute()
            }
        }

        // 6. transit board around the live position
        if let position = fix?.point {
            transport.refresh(center: position, nowMs: nowMs)
            arrivals = transport.allArrivals()
        }
    }
}

File: native/android/app/src/main/java/com/smartmaps/ae/core/AECore.kt
package com.smartmaps.ae.core

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * AECore — the A/E All-In-One orchestrator.
 *
 * Owns every engine and drives one dynamic tick loop. Engines only ever meet
 * here; they never call each other. Modules activate purely by having data —
 * no modes, no configuration, no fallback logic. UI observes the StateFlows;
 * it never reaches into an engine.
 *
 * Tick pipeline (every [tickMs]):
 *   1. movement  ← IMU window (+ GNSS speed as corroboration)
 *   2. fused fix ← inverse-variance fusion of all fresh sources, IMU-gated
 *   3. snap      ← road graph projection with hysteretic lock
 *   4. DR        ← when no fresh fix: advance along the snapped edge
 *   5. route     ← live remaining distance / ETA / next instruction
 *   6. transit   ← provider board refresh around the fused position
 */
class AECore(
    private val cache: OfflineGeometryCache,
    private val scope: CoroutineScope,
    private val tickMs: Long = 200L,
    private val now: () -> Long = System::currentTimeMillis
) {
    val fusion = LocationFusionEngine()
    val movement = MovementEngine()
    val spatial = SpatialEngine()
    val deadReckoning = DeadReckoningEngine()
    val routing = RoutingEngine()
    val transport = TransportTimeEngine()

    // ── observable state (UI binds to these) ────────────────────────────────
    private val _fix = MutableStateFlow<FusedFix?>(null)
    val fix: StateFlow<FusedFix?> = _fix.asStateFlow()

    private val _snap = MutableStateFlow<SnapResult?>(null)
    val snap: StateFlow<SnapResult?> = _snap.asStateFlow()

    private val _movementState = MutableStateFlow(MovementEstimate(MotionState.UNKNOWN, 0.0, 0.0, 0L))
    val movementState: StateFlow<MovementEstimate> = _movementState.asStateFlow()

    private val _route = MutableStateFlow<Route?>(null)
    val route: StateFlow<Route?> = _route.asStateFlow()

    private val _nextStep = MutableStateFlow<RouteStep?>(null)
    val nextStep: StateFlow<RouteStep?> = _nextStep.asStateFlow()

    private val _remaining = MutableStateFlow(Pair(0.0, 0.0)) // (distM, etaS)
    val remaining: StateFlow<Pair<Double, Double>> = _remaining.asStateFlow()

    private val _arrivals = MutableStateFlow<List<TransitArrival>>(emptyList())
    val arrivals: StateFlow<List<TransitArrival>> = _arrivals.asStateFlow()

    private var loop: Job? = null
    private var lastTickMs = 0L
    private var lastGoodFixMs = 0L

    // ── lifecycle ───────────────────────────────────────────────────────────

    /** Load the freshest cached graph (if any) and start the tick loop. */
    fun start(graphKey: String = "default") {
        if (loop != null) return
        cache.load(graphKey)?.let { spatial.setGraph(it) }
        lastTickMs = now()
        loop = scope.launch {
            while (true) {
                tick()
                delay(tickMs)
            }
        }
    }

    fun stop() {
        loop?.cancel()
        loop = null
    }

    /** Install a road graph (e.g. freshly downloaded) — write-through cache. */
    fun installGraph(key: String, graph: RoadGraph) {
        cache.save(key, graph)
        spatial.setGraph(graph)
    }

    // ── sensor feeds (SensorHub pushes here) ────────────────────────────────

    fun pushFix(fix: RawFix) = fusion.ingest(fix)

    fun pushMotion(sample: MotionSample) = movement.push(sample)

    // ── routing ─────────────────────────────────────────────────────────────

    /** Compute and activate a route from the current position (or [fromOverride]). */
    fun navigateTo(destination: GeoPoint, fromOverride: GeoPoint? = null): Route? {
        val from = fromOverride ?: _snap.value?.point ?: _fix.value?.point ?: return null
        val computed = routing.route(spatial.graph(), from, destination)
        _route.value = computed
        return computed
    }

    fun clearRoute() {
        _route.value = null
        _nextStep.value = null
        _remaining.value = 0.0 to 0.0
    }

    // ── the tick ────────────────────────────────────────────────────────────

    private fun tick() {
        val nowMs = now()
        val dtMs = (nowMs - lastTickMs).coerceIn(1L, 2000L)
        lastTickMs = nowMs

        // 1. movement from IMU (+ GNSS speed corroboration)
        val gnssSpeed = _fix.value?.takeIf { nowMs - it.timestampMs < 3000 }?.speedMps
        val move = movement.update(nowMs, gnssSpeed, dtMs)
        _movementState.value = move

        // 2. fuse all fresh location sources
        val fused = fusion.fuse(nowMs, move, dtMs)

        // 3–4. snap to road geometry, or dead-reckon through the outage
        if (fused != null) {
            val snapped = spatial.snap(fused, dtMs)
            if (fused.sources.isNotEmpty() && fused.timestampMs > lastGoodFixMs) {
                lastGoodFixMs = fused.timestampMs
            }
            val stale = nowMs - lastGoodFixMs > 3000
            if (!stale && snapped != null) {
                _snap.value = snapped
                _fix.value = fused.copy(
                    point = snapped.point,
                    headingDeg = spatial.alignedHeading(fused, snapped)
                )
                deadReckoning.seed(fused.speedMps, fused.headingDeg)
            } else {
                // 4. outage: advance along the last snapped edge, IMU-decayed
                val dr = deadReckoning.advance(spatial, _snap.value, move.imuActivity, dtMs)
                if (dr != null) {
                    _snap.value = dr
                    _fix.value = FusedFix(
                        point = dr.point,
                        accuracyM = (_fix.value?.accuracyM ?: 20.0) + 0.5 * (dtMs / 1000.0),
                        speedMps = deadReckoning.currentSpeedMps(),
                        headingDeg = dr.headingDeg,
                        sources = setOf(FixSource.DEAD_RECKONING),
                        timestampMs = nowMs
                    )
                } else if (fused.sources.isNotEmpty()) {
                    _fix.value = fused
                    if (snapped != null) _snap.value = snapped
                }
            }
        }

        // 5. live route progress
        val activeRoute = _route.value
        val position = _fix.value?.point
        if (activeRoute != null && position != null) {
            _remaining.value = routing.remaining(activeRoute, position)
            _nextStep.value = routing.nextStep(activeRoute, position)
            // Automatic arrival: within 25 m of the end with the route nearly
            // consumed → the route clears itself. Dynamic, no confirm dialogs.
            val end = activeRoute.points.lastOrNull()
            if (end != null && haversineM(position, end) < 25.0 && _remaining.value.first < 40.0) {
                clearRoute()
            }
        }

        // 6. transit board around the live position
        if (position != null) {
            transport.refresh(position, nowMs)
            _arrivals.value = transport.allArrivals()
        }
    }
}
